File: javascript/grep/grep.js
#!/usr/bin/env node

// The above line is a shebang. On Unix-like operating systems, or environments,
// this will allow the script to be run by node, and thus turn this JavaScript
// file into an executable. In other words, to execute this file, you may run
// the following from your terminal:
//
// ./grep.js args
//
// If you don't have a Unix-like operating system or environment, for example
// Windows without WSL, you can use the following inside a window terminal,
// such as cmd.exe:
//
// node grep.js args
//
// Read more about shebangs here: https://en.wikipedia.org/wiki/Shebang_(Unix)

const fs = require('fs');
const path = require('path');

function readLines(file) {
    const data = fs.readFileSync(path.resolve(file), { encoding: 'utf-8' });
    return data.split(/\r?\n/);
}

const VALID_OPTIONS = [
    'n', // add line numbers
    'l', // print file names where pattern is found
    'i', // ignore case
    'v', // reverse files results
    'x', // match entire line
];

const ARGS = process.argv;

const processFile = (file, pattern, flags, searchMultipleFiles) => {

    const regex = new RegExp(
        flags.includes('-x') ? `^${pattern}$` : pattern, 
        flags.includes('-i') ? 'i' : '' );

    let matches = [];

    const predicate = (text) => flags.includes('-v') ? !regex.test(text) : regex.test(text);
    const assemble = (line, lineNumber) => (searchMultipleFiles ? `${file}:` : '') 
        + (flags.includes('-n') ? `${lineNumber}:${line}` : line);

    matches = readLines(file)
        .reduce((acc, line, lineNumber) => 
            predicate(line) ? [...acc, assemble(line, lineNumber + 1)] : acc, []);

    if (flags.includes('-l') && matches.length > 0) {
        return file;
    }

    return matches.join('\n');
};

let args = ARGS.slice(2);

const isFile = (item) => item.match(/\w+\.txt/);
const isFlag = (item) => item.startsWith('-');
const flags = args.filter(isFlag);
const files = args.filter(isFile);
const pattern = args.filter(item => !isFile(item) && !isFlag(item));


let res = files.map(file => 
    processFile(file, pattern, flags, files.length > 1))
               .filter(r => r.length > 0).join('\n');

console.log(res);

File: python/grep/grep.py
import re


def grep(pattern: str, flags: list[str], files: list[str]) -> str:
    return ''.join([search_pattern(file, pattern, flags, len(files) > 1)
                     for file in files])


def search_pattern(file: str, pattern: str,
                   flags: list[str], multiple_file_search: bool) -> str:

    pattern_str = f'^{pattern}$' if '-x' in flags else f'.*{pattern}.*'
    f = re.IGNORECASE if '-i' in flags else 0
    the_pattern = re.compile(pattern_str, f)
    input_str = open(file).read().strip()
    output = ''

    for (index, line) in enumerate(input_str.split('\n')):
        if '-v' in flags:
            if not the_pattern.match(line) and '-v' in flags:
                output += (f'{file}:' if multiple_file_search else '') \
                                + (f'{index + 1}:' if '-n' in flags else '') \
                                + line + '\n'

        elif the_pattern.match(line):
            output += (f'{file}:' if multiple_file_search else '') \
                            + (f'{index + 1}:' if '-n' in flags else '') \
                            + line + '\n'

    if output and '-l' in flags:
        output = file + '\n'

    return output
